### PR TITLE
Refactor QasmConvertible[Gate/Operation] into a magic method protocol

### PR DESCRIPTION
**Pull request by Strilanc**
- Added SupportsQasm[WithArgs][AndQubits]
- Sorted some import lists
- Renamed `known_qasm_output` to just `_qasm_`
- Added `_qasm_` method to `cirq.Circuit`
- Renamed `QasmOutputArgs` to `QasmArgs`
- Deleted `QasmConvertibleGate` and `QasmConvertibleOperation`

Fixes https://github.com/quantumlib/Cirq/issues/1031

### CODE REVIEW

File: cirq/circuits/circuit.py [hunk 1]
```diff
@@ -1342,6 +1342,9 @@ def _resolve_parameters_(self,
                 param_resolver))
         return resolved_circuit
 
+    def _qasm_(self) -> str:
+        return self.to_qasm()
+
     def to_qasm(self,
                 header: str = 'Generated from Cirq',
```
Comment (dabacon) [inline cirq/circuits/circuit.py]: should this pull in the version of Cirq it was generated from?

File: cirq/protocols/qasm.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+TDefault = TypeVar('TDefault')
+
+RaiseTypeErrorIfNotProvided = ([],)  # type: Any
+
+
+class QasmArgs(string.Formatter):
+    """
+    Attributes:
+        precision: The number of digits after the decimal to show for numbers in
+            the text diagram.
```
Comment (dabacon) [inline cirq/protocols/qasm.py]: reference to text diagram is slightly odd, maybe "qasm output"?

**Comment by bryano** [comment]
Is the [docstring](https://github.com/quantumlib/Cirq/blob/master/cirq/circuits/qasm_output.py#L15) of  qasm_output.py right?

Is it documented anywhere which flavor/version of QASM is being targeted?

File: cirq/circuits/circuit.py (hunk 1 above)
Comment (Strilanc) [reply cirq/circuits/circuit.py]: Done.

File: cirq/protocols/qasm.py (hunk 2 above)
Comment (Strilanc) [reply cirq/protocols/qasm.py]: Done.

**Comment by Strilanc** [comment]
Fixed the doc string

**Pull request merged**